python/lsst/ip/diffim/detectAndMeasure.py

```diff
@@ -453,33 +425,61 @@ def _deblend(self, difference, positiveFootprints, negativeFootprints):
             Deblended positive and negative polarity footprints measured on
             ``difference``.
         """
-        def makeFootprints(sources):
-            footprints = afwDetection.FootprintSet(difference.getBBox())
-            footprints.setFootprints([src.getFootprint() for src in sources])
-            return footprints
-
-        def deblend(footprints):
-            """Deblend a positive or negative footprint set,
-            and return the deblended children.
-            """
-            sources = afwTable.SourceCatalog(self.schema)
-            footprints.makeSources(sources)
-            self.deblend.run(exposure=difference, sources=sources)
-            self.setPrimaryFlags.run(sources)
-            children = sources["detect_isDeblendedSource"] == 1
-            sources = sources[children].copy(deep=True)
-            # Clear parents, so that measurement plugins behave correctly.
-            sources['parent'] = 0
-            return sources.copy(deep=True)
-
-        positives = deblend(positiveFootprints)
-        negatives = deblend(negativeFootprints)
-
-        sources = afwTable.SourceCatalog(self.schema)
-        sources.reserve(len(positives) + len(negatives))
-        sources.extend(positives, deep=True)
-        sources.extend(negatives, deep=True)
-        return sources, makeFootprints(positives), makeFootprints(negatives)
+        # Merge the positive and negative footprints, but don't grow them.
+        merged_footprints = positiveFootprints
+        merged_footprints.merge(negativeFootprints, 0, 0, False)
+
+        # Create a source catalog from the footprints.
+        table = afwTable.SourceTable.make(self.schema, idFactory)
+        sources = afwTable.SourceCatalog(table)
+        merged_footprints.makeSources(sources)
+
+        # Sky sources must be added before deblending, otherwise the
+        # sources with parent == 0 will be out of order and SFM cannot run.
```

Comment:
I'll change to "measurement tasks." The point is that measurement plugins require all of the children with the same parent ID to be contiguous, but the old version of this task added the sky sources after deblending, so if deblending had worked as expected then measurement would fail.

Reply:
I am a bit surprised by that, because I think that deblending was working on "isolated" positive blends before (it was just messing up other things), and we didn't see failures.

Reply:
Yeah, it's actually because before it wasn't doing any deblending, so all of the sources had `parent==0`. The measurement framework has a call to `SourceCatalog.getChildren(0)`, where the argument is the parent ID. This is done to run all of the measurements on the parents, but in the new catalog if the sky objects were appended to it then the records with parent == 0 was discontinuous.